openvdb/points/PointMove.h

```diff
@@ -0,0 +1,997 @@
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// IN NO EVENT SHALL THE COPYRIGHT HOLDERS' AND CONTRIBUTORS' AGGREGATE
+// LIABILITY FOR ALL CLAIMS REGARDLESS OF THEIR BASIS EXCEED US$250.00.
+//
+///////////////////////////////////////////////////////////////////////////
+//
+/// @author Dan Bailey
+///
+/// @file PointMove.h
+///
+/// @brief Ability to move VDB Points using a custom deformer.
+///
+/// Deformers used when moving points are in world space by default and must adhere
+/// to the interface described in the example below:
+/// @code
+/// struct MyDeformer
+/// {
+///     // A reset is performed on each leaf in turn before the points in that leaf are
+///     // deformed. A leaf and depth-first index within the tree are supplied as the
+///     // arguments, which matches the functor interface for LeafManager::foreach().
+///     template <typename LeafNoteType>
+///     void reset(LeafNoteType& leaf, size_t idx);
+///
+///     // Evaluate the deformer and modify the given position to generate the deformed
+///     // position. An index iterator is supplied as the argument to allow querying the
+///     // point offset or containing voxel coordinate.
+///     template <typename IndexIterT>
+///     void apply(Vec3d& position, const IndexIterT& iter) const;
+/// };
+/// @endcode
+///
+/// @note The DeformerTraits struct (defined in PointMask.h) can be used to configure
+/// a deformer to evaluate in index space.
+
+
+#ifndef OPENVDB_POINTS_POINT_MOVE_HAS_BEEN_INCLUDED
+#define OPENVDB_POINTS_POINT_MOVE_HAS_BEEN_INCLUDED
+
+#include <openvdb/openvdb.h>
+
+#include <openvdb/points/PointDataGrid.h>
+#include <openvdb/points/PointMask.h>
+
+#include <tbb/concurrent_vector.h>
+
+#include <vector>
+#include <map>
+#include <unordered_map>
+
+class TestPointMove;
+
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace points {
+
+// dummy object for future use
+namespace future { struct Advect { }; }
+
+
+/// @brief Move points in a PointDataGrid using a custom deformer
+/// @param points           the PointDataGrid containing the points to be moved.
+/// @param deformer         a custom deformer that defines how to move the points.
+/// @param filter           an optional index filter
+/// @param objectNotInUse   for future use, this object is currently ignored
+/// @param threaded         enable or disable threading  (threading is enabled by default)
+template <typename PointDataGridT, typename DeformerT, typename FilterT = NullFilter>
+inline void movePoints(PointDataGridT& points,
+                       DeformerT& deformer,
+                       const FilterT& filter = NullFilter(),
+                       future::Advect* objectNotInUse = nullptr,
```

Comment:
can you give some insight into why this argument is here?

Reply:
@richhones It is a future hook to allow injecting points from other VDB points grids into an existing one while moving the points. This allows for a single re-bucketing operation instead of being required to perform re-bucketing in two separate steps for moving and merging. A convenience API will also be offered to perform merging without requiring a deformer. The main reason this is a generic object pointer and not simply a vector of PointDataGrid pointers is to be able to optionally filter these secondary point grids while merging. Clearly, the future hook eliminates the need to make a breaking change to the API and deprecate and support the old function for a time, however as merging is a new and quite distinct behaviour, I didn't want it to hold up the introduction of point moving.